### PR TITLE
[Feature](bangc-ops): new op roi_pooling_backward

### PR DESCRIPTION
**Pull request by dcmee336**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new op roi_pooling_backward.

## 2. Modification
create bangc-ops/kernels/roi_pooling_backward/roi_pooling_backward.cpp
create bangc-ops//test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.cpp
create bangc-ops//test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.h
create bangc-ops//test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/test_case/case_0.prototxt

change bangc-ops/mlu_op.h
change docs/bangc-docs/user_guide/9_operators/index.rst
change bangc-ops/kernels/kernel_wrapper/lib/libextops.a
change bangc-ops/kernels/kernel_wrapper/wrapper.h

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          U1       |
|        3       |Layouts                               |          grads, grads_image, argmax: NHWC   rois: ARRAY     |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float          |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/                   |ALL PASS          |          |
|Mult-tensor test     |Supports 1-8 dims                 |NOT SUPPORT          |          |
|Layout test          |Supports NHWC                | ALL PASS         |          |
|Zero element test    |Whether to support this test      |   ALL PASS       |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|    ALL PASS      |          |
|mlu-only mode test   |--mlu-only,see [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)|  ALL PASS        |          |
|Mult-platform test   |MLU370/MLU590                     | ALL PASS         |          |
|Nan/INF test         |Whether to support this test      |  ALL PASS        |          |
|Memory leak check    |Test result                       |    ALL PASS      |          |
|Code coverage check  |Test result                       |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|roi_pooling_backward|  8  |  198.79  |     |    | 0   |float    |grads: [1, 2, 2, 2], grads_image: [1, 6, 6, 2]     |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|roi_pooling_backward|6    | 29301.7   |    |    |  0  | float   | grads: [1, 2, 2, 2], grads_image: [1, 6, 6, 2]     |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
                        const mluOpTensorDescriptor_t output_desc,
                        void *output,
                        int *argmax);
+
+// Group:RoiPoolingBackward
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13877]: 用户手册不更新么？

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+/*!
+ * @brief Computes the gradients of image \b grads_image based on the gradients \b grads and
+ * region proposals \b rois to perform the backpropagation of :: mluOpRoiPoolingBackward operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13880]: 这里没看懂，这个op是做什么的呀？怎么用它解释它自己？

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * The descriptor of the region proposals tensor. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] rois
+ * Pointer to the MLU memory that stores the region proposals tensor.
+ * @param[in] argmax_desc
+ * The descriptor of the argmax tensor that stores the index returned by
+ * ::mluOpRoiPoolingForward. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] argmax
+ * Pointer to the MLU memory that stores the argmax tensor.
+ * @param[in] spatial_scale
+ * A scaling factor that specifies how to map the ROIs coordinates in the origin image to
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13902]:
```suggestion
 * A scaling factor that specifies how to map the ROIs coordinates in the original image to
```

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * - The value of \b pooling_mode only supports \p MLUOP_POOLING_MAX.
+ * - The grads tensor, argmax tensor and grads_image tensor must have four dimensions.
+ * - The size of the lowest dimension of grads tensors, argmax tensor and grads_image tensor must be the same.
+ * - The size of each dimension of grads tensor and argmax tensor must be the same.
+ * - The total number of dimensions of ROIs tensor must be 2.
+ * - The size of the highest dimension of grads tensor, rois tensor and argmax tensor must be the same.
+ * - The shape of \b rois must be [rois_num, 5].
+ * - \b Rois[i] consists of [rois_num, x1, y1, x2, y2]. \b rois_num specifies which image this ROI is in,
+ * and should be in the range of [0, batch_num - 1]. \b x1 and \b y1 specify the starting coordinate of the ROI
+ * in origin image. \b x2 and \b y2 specify the ending coordinate of this ROI in origin image. \b x1 and \b y1
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13938]:
```suggestion
 * in the original image. \b x2 and \b y2 specify the ending coordinate of this ROI in the original image. \b x1 and \b y1
```

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * @param[in] argmax_desc
+ * The descriptor of the argmax tensor that stores the index returned by
+ * ::mluOpRoiPoolingForward. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] argmax
+ * Pointer to the MLU memory that stores the argmax tensor.
+ * @param[in] spatial_scale
+ * A scaling factor that specifies how to map the ROIs coordinates in the origin image to
+ * the coordinates in the output.
+ * @param[in] grads_image_desc
+ * The descriptor of the gradients tensor of original images. For detailed information, see ::mluOpTensorDescriptor_t.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13905]:
```suggestion
 * The descriptor of the gradients tensor of the original images. For detailed information, see ::mluOpTensorDescriptor_t.
```

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ *   - argmax tensor: \b MLUOP_LAYOUT_NHWC.
+ *   - grads_image tensor: \b MLUOP_LAYOUT_NHWC.
+ *
+ * @par Scale Limitation
+ * - The value of \b pooling_mode only supports \p MLUOP_POOLING_MAX.
+ * - The grads tensor, argmax tensor and grads_image tensor must have four dimensions.
+ * - The size of the lowest dimension of grads tensors, argmax tensor and grads_image tensor must be the same.
+ * - The size of each dimension of grads tensor and argmax tensor must be the same.
+ * - The total number of dimensions of ROIs tensor must be 2.
+ * - The size of the highest dimension of grads tensor, rois tensor and argmax tensor must be the same.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13934]: 统一一下格式 参数都加 \b

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * - The shape of \b rois must be [rois_num, 5].
+ * - \b Rois[i] consists of [rois_num, x1, y1, x2, y2]. \b rois_num specifies which image this ROI is in,
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13936]:
```suggestion
 * - \b rois[i] consists of [rois_num, x1, y1, x2, y2]. \b rois_num specifies which image this ROI is in,
```

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * - \b Spatial_scale should be in the range of (0, 1].
+ * - The value of argmax tensors with the data type is int32_t and should be in the range of [0, \f$2^(31)-1\f$].
+ *
+ * @par Note
+ * - In general, set the values of \b argmax according to the result returned by ::mluOpRoiPoolingForward.
+ * Otherwise, these values may be regarded as invalid and will not be used in this operation.
+ * - When \b rois contains NaN or infinity, it may cause undefined behavior.
+ *
+ * @par Example
+ * - The example of the roi_pooling_backward operation is as follows:
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:13949]: 这行是不是可以删掉

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h:13877]: userguide已更新

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 写错了，应该是RoiPoolingForward的反向，已更改

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 8 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已删除

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 9]
```diff
@@ -761,6 +761,12 @@ mluOpRoiPoolingForward
 ----------------------------------
 该算子用于目标检测模型，在经过卷积层计算后的feature map上，针对检测重点关注的区域，即不同ROI对应的feature map区域进行池化，以得到相同规模的输出，进行全连接计算，满足整个网络训练任务。
 
+.. _roi_pooling_backward:
+
+mluOpRoiPoolingBackward
+----------------------------------
+该算子为 ``mluOpRoiPoolingForward`` 的反向算子，反向传播中损失函数对输入层节点的梯度，为各个有可能的候选区域中输出梯度的累加。
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:768]:
```suggestion
该算子为  :ref:`roi_pooling_forward` 的反向算子，反向传播中损失函数对输入层节点的梯度，为各个有可能的候选区域中输出梯度的累加。
```

**Review by AquamarineSu — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 9 above)
Comment (dcmee336) [reply docs/bangc-docs/user_guide/9_operators/index.rst]: 已修改

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
                        const mluOpTensorDescriptor_t output_desc,
                        void *output,
                        int *argmax);
+
+// Group:RoiPoolingBackward
+/*!
+ * @brief Computes the gradients of image \b grads_image based on the gradients \b grads and
+ * region proposals \b rois to perform the backpropagation of :: mluOpRoiPoolingForward operation.
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:13880]: 两个冒号多了个空格在后面

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * A scaling factor that specifies how to map the ROIs coordinates in the original image to
+ * the coordinates in the output.
+ * @param[in] grads_image_desc
+ * The descriptor of the gradients tensor of the original images. For detailed information, see
+ * ::mluOpTensorDescriptor_t.
+ * @param[out] grads_image
+ * Pointer to the MLU memory that stores the gradients tensor of the original images.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH.
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:13911]: 少一个MLUOP_INTERNAL_ERROR

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * @par Data Layout
+ * - The supported data layout of \b grads, \b rois, \b argmax, \b grads_image are as follows:
+ *   - input tensor: \b MLUOP_LAYOUT_NHWC.
+ *   - ROIs tensor: \b MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
+ *   - argmax tensor: \b MLUOP_LAYOUT_NHWC.
+ *   - grads_image tensor: \b MLUOP_LAYOUT_NHWC.
+ *
+ * @par Scale Limitation
+ * - The value of \b pooling_mode only supports \p MLUOP_POOLING_MAX.
+ * - The \b grads tensor, \b argmax tensor and \b grads_image tensor must have four dimensions.
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:13931]: must be 4-D tensor感觉比较好

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * - The size of the lowest dimension of \b grads tensors, \b argmax tensor and \b grads_image tensor must be the same.
+ * - The size of each dimension of \b grads tensor and \b argmax tensor must be the same.
+ * - The total number of dimensions of \b rois tensor must be 2.
```
[the remark this answers is not in the document]
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改为The \b rois tensor must be 2-D tensor.

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已添加

File: bangc-ops/mlu_op.h (hunk 10 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * ::mluOpTensorDescriptor_t.
+ * @param[out] grads_image
+ * Pointer to the MLU memory that stores the gradients tensor of the original images.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH, ::MLUOP_INTERNAL_ERROR.
+ *
+ * @par Data Type
+ * - This function supports the following data types for gradient tensor \b grads, region proposals tensor \b rois,
+ *   argmax_tensor \b argmax and output tensor \b grads_image. The data type of \b grads, \b rois, and \b grads_image
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:13915]:
```suggestion
 *   argmax_tensor \b argmax and output tensor \b grads_image. The data types of \b grads, \b rois, and \b grads_image
```

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * @par Scale Limitation
+ * - The value of \b pooling_mode only supports \p MLUOP_POOLING_MAX.
+ * - The \b grads tensor, \b argmax tensor and \b grads_image tensor must be 4-D tensor.
+ * - The size of the lowest dimension of \b grads tensors, \b argmax tensor and \b grads_image tensor must be the same.
+ * - The size of each dimension of \b grads tensor and \b argmax tensor must be the same.
+ * - The \b rois tensor must be 2-D tensor.
+ * - The size of the highest dimension of \b grads tensor, \b rois tensor and \b argmax tensor must be the same.
+ * - The shape of \b rois must be [rois_num, 5].
+ * - \b rois[i] consists of [rois_num, x1, y1, x2, y2]. \b rois_num specifies which image this ROI is in,
+ * and should be in the range of [0, batch_num - 1]. \b x1 and \b y1 specify the starting coordinate of the ROI
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:13938]: 缩进

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -13873,6 +13873,108 @@ mluOpRoiPoolingForward(mluOpHandle_t handle,
...
+ * in the original image. \b x2 and \b y2 specify the ending coordinate of this ROI in the original image. \b x1 and \b
+ * y1 should be greater than or equal to 0. \b x2 should be greater than \b x1. \b y2 should be greater than \b y1.
+ * - \b Spatial_scale should be in the range of (0, 1].
+ * - The value of argmax tensors with the data type is int32_t and should be in the range of [0, \f$2^(31)-1\f$].
+ *
+ * @par Note
+ * - In general, set the values of \b argmax according to the result returned by ::mluOpRoiPoolingForward.
+ * Otherwise, these values may be regarded as invalid and will not be used in this operation.
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:13946]: 这

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.cpp [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <algorithm>
+#include <iostream>
+#include <vector>
```
Comment (PetrelYy) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.cpp:25]: 这三个有用到吗？  .h 里 #include <vector> 也可以删掉

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 18]
```diff
@@ -761,6 +761,12 @@ mluOpRoiPoolingForward
 ----------------------------------
 该算子用于目标检测模型，在经过卷积层计算后的feature map上，针对检测重点关注的区域，即不同ROI对应的feature map区域进行池化，以得到相同规模的输出，进行全连接计算，满足整个网络训练任务。
 
+.. _roi_pooling_backward:
+
+mluOpRoiPoolingBackward
+----------------------------------
+该算子为 :ref:`roi_pooling_forward` 的反向算子，反向传播中损失函数对输入层节点的梯度，为各个有可能的候选区域中输出梯度的累加。
```
Comment (PetrelYy) [inline docs/bangc-docs/user_guide/9_operators/index.rst:768]: 网页打开，这里不会跳转， 有测试PDF格式是否准确么？ 可以问下文档组

File: bangc-ops/mlu_op.h (hunk 15 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 16 above)
Comment (dcmee336) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.cpp (hunk 17 above)
Comment (dcmee336) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_pooling_backward/roi_pooling_backward.cpp]: 这三个已删除，.h中删掉了一个

**Review by PetrelYy — APPROVED** [review]

**Pull request merged**